### ISSUE TITLE
Issue in generated fortios_system_interface -> allowedaccess

### ISSUE THREAD

**Issue by JonTheNiceGuy** [fortinet-solutions-cse/ansible_generator#6]
Please see generated output in https://github.com/fortinet-solutions-cse/ansible_fgt_modules/blob/master/v6.0.2/system/fortios_system_interface.py#L1781-L1785

Currently, this only permits setting a single item ("one of" ['ping', 'http', 'https', ... etc]) instead of being able to select individual items.

*ideally* this would be an options list:

```
  - name: Configure interfaces.
    fortios_system_interface:
      system_interface:
        state: present
        allowaccess:
          ping: True
          http: False
          https: True
          fgfm: False
        interface: port1
```

Without digging into either the ansible_generator, [fortiosapi](https://github.com/fortinet-solutions-cse/fortiosapi) or the [generated ansible module](https://github.com/fortinet-solutions-cse/ansible_fgt_modules] code, I can't express whether this is the "right way" to do this, or even if it's a "feasible" way to do it.

Also, FAO @tohillm

**Comment by mamunozgonzalez** [comment]
Right @JonTheNiceGuy, there seems to be a bug here. I'd give max prio to this. 
My impression is that we will have to build a list manually with those fields which allow for multiple entries in a single list. This way the generator knows that this is a special case and can generate special code to allow for multiple choices in a single entry. 
I'm trying to fetch that info from the schema but I have the feeling that it won't be available.

Thanks for reporting this. Very good catch!

**Comment by JonTheNiceGuy** [comment]
Yehr, I agree. You almost need an additional type in the schema, perhaps like this:

```
"allowaccess":{
  "name":"allowaccess",
  "category":"unitary",
  "type":"multi-option",
  "help":"Permitted types of management access to this interface.",
  "options":[
    {
      "name":"ping",
      "help":"PING access."
    },
    {
      "name":"https",
      "help":"HTTPS access."
    },
    {
      "name":"ssh",
      "help":"SSH access."
    },
    {
      "name":"snmp",
      "help":"SNMP access."
    },
    {
      "name":"http",
      "help":"HTTP access."
    },
    {
      "name":"telnet",
      "help":"TELNET access."
    },
    {
      "name":"fgfm",
      "help":"FortiManager access."
    },
    {
      "name":"radius-acct",
      "help":"RADIUS accounting access."
    },
    {
      "name":"probe-response",
      "help":"Probe access."
    },
    {
      "name":"capwap",
      "help":"CAPWAP access."
    },
    {
      "name":"ftm",
      "help":"FTM access."
    }
  ]
}
```

I suspect this has other impacts upstream though (and I doubt that will be backported into <6.0.4 :grin: ) Maybe something you can propose into the FGT team?

**Comment by mamunozgonzalez** [comment]
I have a fix working locally to solve this issue. I'll be doing some more extensive tests/regression and I will upload a fix shortly. 

For now it seems that the schema does not have the info to indicate if an attribute is a 'multilist'/'multioption' so we have to put this info in a file before running the generator:

Something like this:
```
 {
     "system_interface":  [ ["allowaccess"],
                            ["ipv6", "ip6-allowaccess"] ]
 }
```

For now I have only identified these two attributes but I know there will be more. So feel free to update the list and push the changes or let me know and I will update it.

In the meantime I keep on investigating if there is a reasonable way to detect these attributes automatically but I'm afraid it will not be possible.

I'll release a fix shortly today.

**Comment by JonTheNiceGuy** [comment]
Once again, amazing support from you @mamunozgonzalez! Thanks!

**Comment by mamunozgonzalez** [comment]
You are very welcome. This also helps us a lot to improve our stuff.

I have run a regression with the latest changes using https://github.com/fortinet-solutions-cse/ansible_generator/blob/master/run_test.sh. All seems to be ok (**Success: 176  Failed: 0**).
 I have also run an example with fortios_system_interface_example.yml (I'll upload it later).

This commit contains the fix: https://github.com/fortinet-solutions-cse/ansible_generator/commit/4cf3df529c8452f37d73b339ed59740e40e71f61

I have also generated the new modules and uploaded them to ansible_fgt_modules:

https://github.com/fortinet-solutions-cse/ansible_fgt_modules/commit/3104d7eb4763b9e296ec5c51c769cc6aa4b1cef6

Btw, I'll do a couple of brief modifications more in the generator to simplify the code (just cosmetic stuff, not affecting performance or features). 

Thanks for the feedback!

**Comment by tohillm** [comment]
@mamunozgonzalez  @JonTheNiceGuy 

Hello Both.
I am using the new 'fortios_system_snmp_user.py' generated from this issue:

fortinet-solutions-cse/ansible_fgt_modules@3104d7e
https://github.com/fortinet-solutions-cse/ansible_fgt_modules/commit/3104d7eb4763b9e296ec5c51c769cc6aa4b1cef6

I then run the following:

`   - name: SNMP user configuration.
     fortios_system_snmp_user:
       host:  "{{ host1 }}"
       username: "{{ username }}"
       password: "{{ password }}"
       vdom:  "{{ vdom }}"
       https: "True"
       system_snmp_user:
         state: "present"
         auth-proto: "md5"
         auth-pwd: "XXX"
         events:
            - cpu-high
            - mem-low
         ha-direct: "enable"
         name: "XXX"
         notify-hosts: "x.x.x.x"
         priv-proto: "aes"
         priv-pwd: "XXX"
         queries: "enable"
         query-port: "161"
         security-level: "auth-priv"
#        source-ip: "84.230.14.43"
         status: "enable"
#        trap-lport: "18"
#        trap-rport: "19"
         trap-status: "enable"
`

I get the following result:
`fatal: [localhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "host": "XXX", 
            "https": true, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "system_snmp_user": {
                "auth-proto": "md5", 
                "auth-pwd": "auth-pass1", 
                "events": "['cpu-high', 'mem-low']", 
                "ha-direct": "enable", 
                "name": "XXX", 
                "notify-hosts": "XXX", 
                "priv-proto": "aes", 
                "priv-pwd": "priv-pass1", 
                "queries": "enable", 
                "query-port": 161, 
                "security-level": "auth-priv", 
                "state": "present", 
                "status": "enable", 
                "trap-status": "enable"
            }, 
            "username": "XXX", 
            "vdom": "root"
        }
    }, 
    "msg": "value of events must be one of: cpu-high, mem-low, log-full, intf-ip, vpn-tun-up, vpn-tun-down, ha-switch, ha-hb-failure, ips-signature, ips-anomaly, av-virus, av-oversize, av-pattern, av-fragmented, fm-if-change, fm-conf-change, bgp-established, bgp-backward-transition, ha-member-up, ha-member-down, ent-conf-change, av-conserve, av-bypass, av-oversize-passed, av-oversize-blocked, ips-pkg-update, ips-fail-open, faz-disconnect, wc-ap-up, wc-ap-down, fswctl-session-up, fswctl-session-down, load-balance-real-server-down, device-new, per-cpu-high, got: ['cpu-high', 'mem-low'] found in system_snmp_user"
}
`
Am I doing something wrong?

**Comment by mamunozgonzalez** [comment]
Hi @tohillm 
 As you know, I included a fix for provisioning multiple values in Ansible modules. However, the information about which attributes can allow for multiple values is not present in the schema, therefore have to provide it manually in this file:

https://github.com/fortinet-solutions-cse/ansible_generator/blob/master/special_attributes.lst

And re-run the generator afterwards.

I have submitted a bug request to FortiGate which will fix the schema issue in next releases but for now we have to modify the special_attributes.lst file manually and re-run the generator.

I see that the attribute that you are trying to use has not been included yet: system_snmp/user/events. Actually we are gradually including this info as we discover new attributes. Thus feel free to modify the file (special_attributes.lst) yourself and submit a Pull Request (for v6.0.2). 

Or let me know and I will update it myself.
